Add proposal for integrating Bridge with Metrics Reporter (#143)
strimzi/proposals: 090-integrate-bridge-with-metrics-reporter.md

Note that this agent depends on the [Kafka JMX Reporter](https://github.com/apache/kafka/blob/3.9.0/clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java) plugin, which is enabled by default.

The only major component not using the `metricsConfig` schema is the [HTTP Bridge](https://github.com/strimzi/strimzi-kafka-bridge), as it directly instantiates the JMX Exporter's `JmxCollector` class with hard coded configuration.
The `KafkaBridge` CRD does not support a type system for metrics configuration, but a simple boolean configuration to enable/disable metrics.

```sh
spec:
  enableMetrics: true
```

## Motivation

We want to support the Strimzi Metrics Reporter as an alternative way of configuring metrics across all components.
In order to provide a consistent user experience, the HTTP Bridge should align with all the other components in terms of metrics configuration. 

## Proposal

As specified in SIP-064, a new metrics config type called `strimziMetricsReporter` will be added to `metricsConfig` schema along with its configuration.

> Only the Strimzi Metrics Reporter's `prometheus.metrics.reporter.allowlist` property will be configurable by users when using the Cluster Operator.

Reviewer: I think it's still confusing referring to the Cluster Operator here. Can we just explain the schema referring to the bridge (maybe mentioning that the same is going to be used for the Cluster Operator)?

This configuration is a comma separated list of regex patterns that is used by the plugin to filter metrics to collect.
All other plugin configuration properties will be locked down, including the port that will be set to 9404 to avoid conflicts with other ports.

```sh
spec:
  kafka:
    metricsConfig:
      type: strimziMetricsReporter
      values:
        allowList:
          - "kafka_log.*"
          - "kafka_network.*"
```

### HTTP Bridge

The HTTP Bridge project will be updated to support multiple types of metrics configurations.

> A new `bridge.metrics` property will be available, and will only accept `jmxPrometheusExporter` and `strimziMetricsReporter`.

Reviewer: ```suggestion
A new `bridge.metrics` property will be available within the application.properties configuration file and will only accept `jmxPrometheusExporter` and `strimziMetricsReporter`.
```

> A new `bridge.metrics` property will be available, and will only accept `jmxPrometheusExporter` and `strimziMetricsReporter`.

Reviewer: I would also expect that not defining it just defaults to the current behaviour by using JMX right? Should we state it here clearly?

Author: Current default is `KAFKA_BRIDGE_METRICS_ENABLED=false`, so when `bridge.metrics` will not be set, then the metrics endpoint will be disabled. This behavior is not changing.

Reviewer: but we are deprecating the env var so it means that if you don't change the logic to default right now we have to do that in the future when the env var is not there anymore?

Reviewer: Just to clarify what I mean ...
The implementation code will take care of KAFKA_BRIDGE_METRICS_ENABLED=false to understand that no metrics have to be exposed. But at some point the code should take care of "the bridge.metrics field is NOT set" to have the same (no metrics). So we should plan to switch to the second check when the env var is removed.

Author: This is already covered by the draft implementation. Let's discuss that when we get to it.

The following example will be provided as a comment in the default properties file.

```sh
bridge.metrics=strimziMetricsReporter
kafka.metric.reporters=io.strimzi.kafka.metrics.KafkaPrometheusMetricsReporter
kafka.prometheus.metrics.reporter.listener.enable=false
kafka.prometheus.metrics.reporter.allowlist=.*
```

The `KAFKA_BRIDGE_METRICS_ENABLED` env var will be deprecated and removed in a future release.
When set, the user will get a warning suggesting to use the new `bridge.metrics` configuration.
In case they are both set, the `bridge.metrics` configuration will take precedence over `KAFKA_BRIDGE_METRICS_ENABLED`.

The `MetricsReporter` class will be updated to also include a new `StrimziCollectorRegistry` that will work similarly to the `JmxCollectorRegistry`.
The `StrimziCollectorRegistry` will include a reference to `PrometheusRegistry.defaultRegistry`, which is the same instance used by the Strimzi Metrics Reporter to collect metrics.

The `kafka_bridge_config_generator.sh` script is used to generate the image configuration based on environment variables.
This script will be updated to also include `bridge.metrics` and related configurations (see the following section).

> The Bridge will also try to load the JMX Exporter configuration file from the path specified by the `bridge.metrics.jmx.exporter.config.path` property.

Reviewer: This sounds to be an addition to the enablement of the Strimzi metrics reporter.
I mean, we could add the Strimzi metrics reporter and leaving the current JMX reporter still working with the hard-coded configuration. I am not against it but if we want to have both things we should make it clearer that this is a new addition and not needed for the Strimzi metrics reporter which is the main goal of the proposal.

Author: Initially, this wasn't part of the proposal, but then I realized that the JMX Exporter configuration is required when using the shared `metricsConfig` schema, and we have it in all the metrics examples. Additionally, I think it adds a useful feature for the standalone Bridge users. Wdyt?

Reviewer: I am fine with it, but let's make it clear it's an addition not something needed for the Strimzi metrics reporter support.

Reviewer: Anyway when you said "it's required" ... don't we have any way to fall back using the default "hard coded" configuration? I think it's mentioned somewhere else in the proposal, or?

Author: > let's make it clear it's an addition not something needed for the Strimzi metrics reporter support

Ok, will do.

> don't we have any way to fall back using the default "hard coded" configuration? 

Correct. Let me fix.

If not specified or not found, it will fall back to the hard coded configuration.

> ### Cluster Operator

Reviewer: I guess this title should reflect what the Strimzi cluster operator will do to configure the bridge. Can we use a better title?

Like for all the other major components, the `KafkaBridge` CRD will also include `metricsConfig` with support for `strimziMetricsReporter` type.

Three new environment variables will be introduced to pass the metrics configuration in the HTTP Bridge container:

1. `STRIMZI_METRICS`: This will contain the `metricsConfig` types to enable (one of `jmxPrometheusExporter` and `strimziMetricsReporter`).
2. `KAFKA_BRIDGE_METRICS_JMX_CONFIG`: Used with `STRIMZI_METRICS=jmxPrometheusExporter` to pass the configuration file path.
3. `KAFKA_BRIDGE_METRICS_SMR_CONFIG`: Used with `STRIMZI_METRICS=strimziMetricsReporter` to pass the plugin configuration.

The `enableMetrics` property will be deprecated and removed in a future release.
In case they are both set, `metricsConfig` will take precedence over `enableMetrics`.

All Strimzi Metrics Reporter configurations will be locked down with the exception of `allowList`.

> Unlike the other major components, the metrics endpoint port will be set to 8080, which is the one already used to expose the REST API.

Reviewer: the metrics endpoint should be disabled not set to 8080 otherwise it will conflict with the bridge one, or?

Author: Yes, I just expressed this poorly. Fixing.

# Integrate Bridge with Metrics Reporter

In [SIP-064](https://github.com/strimzi/proposals/blob/main/064-prometheus-metrics-reporter.md) we introduced the [Strimzi Metrics Reporter](https://github.com/strimzi/metrics-reporter).
This is a Kafka MetricsReporter plugin that directly exposes metrics in Prometheus format via an HTTP endpoint.

## Current situation

The HTTP Bridge allows users to enable or disable the metrics endpoint using the `KAFKA_BRIDGE_METRICS_ENABLED` environment variable.

> When this variable is set to true, the Bridge creates the JMX Exporter's JmxCollector class with hard coded configuration.

Author: Sure.

This configuration is a JMX metrics exporter YAML distributed as an embedded resource within the Bridge's JAR file.

When deploying the Bridge through the Cluster Operator, a similar configuration is available in the KafkaBridge's CRD:

```sh
spec:
  enableMetrics: true
```

This is different from how the metrics endpoint is enabled in all the other major components deployed by the Cluster Operator.
The Cluster Operator exposes metrics through the [Prometheus JMX Exporter](https://github.com/prometheus/jmx_exporter), which can be configured using the shared `metricsConfig` schema.
This schema has a type property that only allows the `jmxPrometheusExporter` value, and a reference to a ConfigMap containing its configuration:

```sh
metricsConfig:
  type: jmxPrometheusExporter
  valueFrom:
    configMapKeyRef:

>       name: $COMPONENT_NAME-metrics

Reviewer: maybe a note to highlight what `$COMPONENT_NAME` means here and possible values?

Author: Yes, this is confusing. Let me fix it.

      key: metrics-config.yml
```

At runtime, the above configuration enables a Java agent which exposes Kafka JMX metrics in Prometheus format through an HTTP endpoint on port 9404.

> Note that this agent depends on the [Kafka JMX Reporter](https://github.com/apache/kafka/blob/3.9.0/clients/src/main/java/org/apache/kafka/common/metrics/JmxReporter.java) plugin, which is enabled by default.

Author: "Requires" should be better.

## Proposal

The Bridge will support both JMX Prometheus Exporter and Strimzi Metrics Reporter as metrics configuration types.
It will also provide new configurations for the integration with Cluster Operator.

### Metrics configuration types

The Bridge project will be updated to support multiple metrics configurations types.
A new `bridge.metrics` property will be available within the application.properties configuration file, and will only accept `jmxPrometheusExporter` and `strimziMetricsReporter`.
Any other value will raise an error and the application will fail to start with an appropriate error message.

When running in standalone mode with `strimziMetricsReporter`, the user will be able to configure any reporter property using the "kafka." prefix.
The following example will be provided as a comment in the default application.properties file:

```sh
bridge.metrics=strimziMetricsReporter
kafka.metric.reporters=io.strimzi.kafka.metrics.KafkaPrometheusMetricsReporter
kafka.prometheus.metrics.reporter.listener.enable=false
kafka.prometheus.metrics.reporter.allowlist=.*
```

> The `KAFKA_BRIDGE_METRICS_ENABLED` environment variable will be deprecated and removed in a future release.

Reviewer: should we make a clearer plan. I.e. deprecated in 0.32.0 (next release when we have the Strimzi metrics reporter) and removed in 0.33.0 or 0.34.0? Of course the releases can be shifted if we don't get this proposal implemented on time for the 0.32.0 release.

Author: Ok, I will propose a schedule so that we can discuss.

When set, the user will get a warning suggesting to use the `bridge.metrics` property.
In case they are both set, `bridge.metrics` will take precedence over `KAFKA_BRIDGE_METRICS_ENABLED`.

The MetricsReporter class will be updated to also include a new StrimziCollectorRegistry that will work similarly to the JmxCollectorRegistry.
The StrimziCollectorRegistry will include a reference to PrometheusRegistry.defaultRegistry, which is the same instance used by the Strimzi Metrics Reporter to collect metrics.

The kafka_bridge_config_generator.sh script is used to generate the image configuration based on environment variables.
This script will be also updated to include `bridge.metrics` and related configurations (see the following section).

The Bridge will try to load the JMX Prometheus Exporter configuration file from the path specified by the `bridge.metrics.jmx.exporter.config.path` property.
If the property is not specified or the file is not found, the Bridge will fall back to the hard coded configuration.
This feature is not strictly required to support the Strimzi Metrics Reporter, but will be used by the Cluster Operator.

> ### Cluster Operator integration

Author: I will add a dedicated section. For `auto.include.jmx.reporter` there is no mention in the Strimzi Metrics Reporter proposal, so we will leave that to the user. Additionally, that property will be removed in Kafka 4.0, and the user would need to be explicitly add the JMX metrics plugin when setting any custom plugin.

> ### Cluster Operator integration

Reviewer: But we could end in a Strimzi version with Kafka 3.9 support, so the `auto.include.jmx.reporter` would be still needed, or?

> ### Cluster Operator integration

Author: Given it's deprecated and not required to enable the Strimzi Metrics Reporter, I would leave this configuration to the user through `.spec.kafka.config`. Wdyt?

## Proposal

The Bridge will support both *Exporter* and *Reporter* as metrics configuration types.
It will also provide new configurations for the integration with Cluster Operator.

### Metrics configuration types

The Bridge project will be updated to support multiple metrics configurations types.
A new `bridge.metrics` property will be available within the application.properties configuration file, and will only accept `jmxPrometheusExporter` and `strimziMetricsReporter`.
Any other value will raise an error and the application will fail to start with an appropriate error message.

When running in standalone mode with `strimziMetricsReporter`, the user will be able to configure any reporter property using the "kafka." prefix.
The following example will be documented and also provided as a comment in the default application.properties file:

```sh
bridge.metrics=strimziMetricsReporter
kafka.metric.reporters=io.strimzi.kafka.metrics.KafkaPrometheusMetricsReporter
kafka.prometheus.metrics.reporter.listener.enable=false
kafka.prometheus.metrics.reporter.allowlist=.*
```

> The MetricsReporter class will be updated to also include a new StrimziCollectorRegistry that will work similarly to the JmxCollectorRegistry.

Author: Sure. Creating a collector registry for each metrics type ensure better isolation, and makes it easier to add or remove metrics types.

The StrimziCollectorRegistry will include a reference to PrometheusRegistry.defaultRegistry, which is the same instance used by the *Reporter* to collect metrics.

The kafka_bridge_config_generator.sh script is used to generate the image configuration based on environment variables.
This script will be also updated to include `bridge.metrics` and related configurations (see the following section).

> The Bridge will try to load the Exporter's configuration file from the path specified by the `bridge.metrics.jmx.exporter.config.path` property.

Reviewer 2: ```suggestion
The Bridge will try to load the configuration file of the *Exporter* from the path specified by the `bridge.metrics.jmx.exporter.config.path` property.
```

Reviewer 2: I'm not clear on this property and the context here? How is it specified and under what circumstances? Is it specified in application.properties if we choose jmxPrometheusExporter?

Author: This configuration property will be used by the Cluster Operator to pass the user-provided Exporter's configuration file to the Bridge (see `metricsConfig` schema). The same property can also be leveraged by standalone Bridge users.

> If the property is not specified or the file is not found, the Bridge will fall back to the hard coded configuration.

Reviewer 2: ```suggestion
If the property is not specified or the file is not found, the Bridge will fall back to the hard-coded configuration.
```

This feature is not strictly required to support the *Reporter*, but will be used by the Cluster Operator.

### Cluster Operator integration

> The KafkaBridge's CRD will also support `metricsConfig` with the addition of `strimziMetricsReporter` type.

Reviewer 2: ```suggestion
The `KafkaBridge` CRD will also support `metricsConfig` with the addition of `strimziMetricsReporter` type.
```

> This is how the Strimzi Reporter configuration will look like:

Reviewer 2: ```suggestion
This is how the Strimzi *Reporter* configuration will look like:
```

```sh
spec:
  metricsConfig:
    type: strimziMetricsReporter
    values:
      allowList:
        - "kafka_log.*"
        - "kafka_network.*"
```

Three new environment variables will be introduced to pass the metrics configuration to the Bridge's container:

- `STRIMZI_METRICS`: This will contain the `metricsConfig` types to enable (one of `jmxPrometheusExporter` and `strimziMetricsReporter`).
- `KAFKA_BRIDGE_METRICS_JMX_CONFIG`: Used with *Exporter* to pass the configuration file path.
- `KAFKA_BRIDGE_METRICS_SMR_CONFIG`: Used with *Reporter* to pass the plugin configuration.

The *Exporter* configuration file will be stored in a ConfigMap and mounted in the Bridge's container.
The full configuration file passed to the Bridge's container will be `/opt/strimzi/custom-config/metrics-config.yml`.

> Metrics will be exposed through the Bridge's HTTP server, so the *Reporter*'s listener will be disabled.

Reviewer 2: ```suggestion
Metrics will be exposed through the Bridge's HTTP server, so the listener of the *Reporter* will be disabled.
```

Other configurations will be locked down with the exception of the `prometheus.metrics.reporter.allowlist` property.

## Affected/not affected projects

The affected projects are Cluster Operator and Kafka Bridge.

## Compatibility

### Bridge

The `KAFKA_BRIDGE_METRICS_ENABLED` environment variable will be deprecated in 0.32.0 and removed in 0.34.0.
When set, the user will get a warning suggesting to use the `bridge.metrics` property.
In case they are both set, `bridge.metrics` will take precedence over `KAFKA_BRIDGE_METRICS_ENABLED`.

### Cluster Operator

> The `enableMetrics` property in KafkaBridge CRD will be deprecated in 0.46.0 and removed in 0.48.0.

Reviewer 2: ```suggestion
The `enableMetrics` property in `KafkaBridge` CRD will be deprecated in 0.46.0 and removed in 0.48.0.
```

When set, the user will get a warning suggesting to use the `metricsConfig` configuration.
In case they are both set, `metricsConfig` will take precedence over `enableMetrics`.

## Rejected alternatives

Lock down and automate the *Reporter* configurations for the standalone Bridge.

> This was rejected because the Bridge allows user to customize Kafka clients and plugins using the application.properties file, which includes commented examples.

Reviewer 2: ```suggestion
This was rejected because the Bridge allows users to customize Kafka clients and plugins using the application.properties file, which includes commented examples.
```

4. *Reporter* configuration used to configure a comma separated list of regex patterns to filter metrics to collect.

The MetricsReporter class will be updated to also include a new `StrimziCollectorRegistry` that will work similarly to the `JmxCollectorRegistry`.
Creating a collector registry for each metrics type ensure better isolation, and makes it easier to add or remove metrics types.
The `StrimziCollectorRegistry` will include a reference to `PrometheusRegistry.defaultRegistry`, which is the same instance used by the *Reporter* to collect metrics.

The Kafka Bridge will try to load a custom Exporter's configuration file from the path specified by the `bridge.metrics.exporter.config.path` property.
If the property is not specified or the file is not found, the Kafka Bridge will fall back to the hard-coded configuration file.
This feature is not strictly required to support the *Reporter*, but will be used by the Cluster Operator.

### Cluster Operator integration

The `KafkaBridge` CRD will also support `metricsConfig` with the addition of `strimziMetricsReporter` type.
This is how the *Reporter* configuration will look like:

> ```sh

Reviewer 3: ```suggestion
```yaml
```

# Integrate Kafka Bridge with Metrics Reporter

> In [SIP-064](https://github.com/strimzi/proposals/blob/main/064-prometheus-metrics-reporter.md), we introduced the [Strimzi Metrics Reporter](https://github.com/strimzi/metrics-reporter).

Reviewer 4: can we use just use `Strimzi Proposal 064`?

### Metrics configuration types

The Kafka Bridge project will be updated to support two metrics configuration types.
A new `bridge.metrics` property will be available within the `application.properties` configuration file, and will only accept `jmxPrometheusExporter` and `strimziMetricsReporter`.
Any other value will raise an error and the application will fail to start with an appropriate error message.

When running in standalone mode with `strimziMetricsReporter`, the user will be able to configure any reporter property using the "kafka." prefix.
The following example will be documented and included, with comments, in the default `application.properties` file.
The `kafka.` prefix is used by the Bridge to pass the following configuration to the internal Kafka clients.

```sh
# uncomment the following lines to enable Strimzi Reporter metrics, check the documentation for more details
bridge.metrics=strimziMetricsReporter #1
kafka.metric.reporters=io.strimzi.kafka.metrics.KafkaPrometheusMetricsReporter #2
kafka.prometheus.metrics.reporter.listener.enable=false #3
kafka.prometheus.metrics.reporter.allowlist=.* #4

> ```

Reviewer: I guess this part of the properties file is an example when the bridge is used on bare metal/vm. In such a case, it's the user to provide the entire configuration (their own application.properties file). The bridge itself doesn't never change the config file on startup to add "something" to it.
One thing we could imagine is that the bridge doesn't update the file but add these value at runtime (in memory) when configuring the internal Kafka clients. My only doubt is that it won't be reflected in what the user configured so the user doesn't know about that.

> ```

Author: We can use the same logic that we have in the Operator, with locked down Reporter configuration and a new `bridge.metrics.allowList` property.

> ```

Reviewer 3: Maybe I missed some of the arguments you are trying to make and maybe my comment was not clear, sorry. It of course does not matter when the operator configures it. But when I configure it myself, why do I need to set both
```
bridge.metrics=strimziMetricsReporter #1
kafka.metric.reporters=io.strimzi.kafka.metrics.KafkaPrometheusMetricsReporter #2
```

They seem to do the same - enable the Strimzi metrics reporter, or? So what is the difference between them?

> ```

Author: The `bridge.metrics` property is the analogous of the `metricsConfig.type` field we have in the CO schemas. Maybe the property name is confusing, but it was inspired by the already existing `bridge.tracing`. It will be used to initialize the Bridge's `MetricsRporter` with the right `CollectorRegistry`.

The `kafka.metric.reporters` property is basically the Kafka `metric.reporters` with the `kafka.` prefix used by the Bridge to group broker configurations.

> ```

Reviewer 3: But why do we need the user to configure both? You say it is `analogous of the metricsConfig.type field` -> but there you configure only the `metricsConfig.type` field.

1. Kafka Bridge configuration used to set the metrics type.
2. Kafka configuration used to set the Kafka metrics reporter implementation.
3. *Reporter* configuration used to disable the HTTP listener (metrics are exposed through the Kafka Bridge listener).
4. *Reporter* configuration used to configure a comma separated list of regex patterns to filter metrics to collect.

The MetricsReporter class will be updated to also include a new `StrimziCollectorRegistry` that will work similarly to the `JmxCollectorRegistry`.
Creating a collector registry for each metrics type ensure better isolation, and makes it easier to add or remove metrics types.
The `StrimziCollectorRegistry` will include a reference to `PrometheusRegistry.defaultRegistry`, which is the same instance used by the *Reporter* to collect metrics.

The `kafka_bridge_config_generator.sh` script is used to generate the Kafka Bridge image configuration based on environment variables.

> This script will be also updated to include `bridge.metrics` and related configurations, as described in the following integration section.

Reviewer: Yeah. This proposal was written down before I started the work on moving bridge configuration. I think there is some changes needed here @fvaleri based on that PR.

> This script will be also updated to include `bridge.metrics` and related configurations, as described in the following integration section.

Author: Ok, but I will wait for that PR to settle down.

> This script will be also updated to include `bridge.metrics` and related configurations, as described in the following integration section.

Reviewer 3: Yeah, I think that makes sense.

The Kafka Bridge will try to load a custom Exporter's configuration file from the path specified by the `bridge.metrics.jmx.exporter.config.path` property.
If the property is not specified or the file is not found, the Kafka Bridge will fall back to the hard-coded configuration file.
This feature is not strictly required to support the *Reporter*, but will be used by the Cluster Operator.

### Cluster Operator integration

The `KafkaBridge` CRD will also support `metricsConfig` with the addition of `strimziMetricsReporter` type.
This is how the *Reporter* configuration will look like:

```sh
spec:
  metricsConfig:
    type: strimziMetricsReporter
    values:
      allowList:
        - "kafka_log.*"

>         - "kafka_network.*"

Author: Yes, the default is `.*`. Example updated.

>         - "kafka_network.*"

Reviewer 3: That is not really what I meant. And it relates to the PR from @OwenCorrigan76 as well. I think we should choose one of the ways here:
1. Make the allowList required
2. Have a good default for them for each component. (And I guess `.*` is not the right default) This would be hardcoded in the code base and change as we for example update our dashabords.

Having a flat default `.*` everywhere is IMHO a bad option, because historically our experience is that users want to limit the metrics exporter:
* For performance reasons in Kafka (this might improve with our own reporter, but I expect it would still be an issue at some amount of metrics)
* For performance reasons in Prometheus (the metrics can be dropped by Prometheus after scraping, but that is inefficient and harder to configure)

>         - "kafka_network.*"

Author: Bridge defaults added. We will address the other components in the related PRs.

```

Three new environment variables will be introduced to pass the metrics configuration to the Bridge's container:

- `STRIMZI_METRICS`: This will contain the `metricsConfig` types to enable (one of `jmxPrometheusExporter` and `strimziMetricsReporter`).
- `KAFKA_BRIDGE_METRICS_JMX_CONFIG`: Used with *Exporter* to pass the configuration file path.
- `KAFKA_BRIDGE_METRICS_SMR_CONFIG`: Used with *Reporter* to pass the plugin configuration.

The *Exporter* user provided configuration file will be stored in a ConfigMap and mounted in the Bridge's container.
The file path passed to the Bridge's container using the `bridge.metrics.jmx.exporter.config.path` property will be `/opt/strimzi/custom-config/metrics-config.yml`.

Metrics will be exposed through the Kafka Bridge's HTTP server, so the listener of the *Reporter* will be disabled.
Other configurations will be locked down with the exception of the `prometheus.metrics.reporter.allowlist` property.

## Affected/not affected projects

The affected projects are Cluster Operator and Kafka Bridge.

## Compatibility

### Bridge

> The `KAFKA_BRIDGE_METRICS_ENABLED` environment variable will be deprecated in 0.32.0 and removed in 0.34.0.

Author: We usually use versions, but I'm fine with dates if you think it's better. Do you mean to deprecate next release and remove in January 2026?

> The `KAFKA_BRIDGE_METRICS_ENABLED` environment variable will be deprecated in 0.32.0 and removed in 0.34.0.

Reviewer 3: We do - but the Bridge minor releases do not have the same set cadence as the operator. The BRidge already used time based definition for example in https://github.com/strimzi/proposals/blob/main/071-deprecate-bridge-openapi-2.md

> The `KAFKA_BRIDGE_METRICS_ENABLED` environment variable will be deprecated in 0.32.0 and removed in 0.34.0.

Author: Ok, makes sense. Will update the proposal.

When set, the user will get a warning suggesting to use the `bridge.metrics` property.
In case they are both set, `bridge.metrics` will take precedence over `KAFKA_BRIDGE_METRICS_ENABLED`.

### Cluster Operator

> The `enableMetrics` property in `KafkaBridge` CRD will be deprecated in 0.46.0 and removed in 0.48.0.

Author: > I guess we can also keep it working until removed in v1 API easily to simplify it for the users.

Ok, there is no reason to remove it before v1. Updated.

> explained how would the migration work as well

What do you mean exactly with migration?

> The `enableMetrics` property in `KafkaBridge` CRD will be deprecated in 0.46.0 and removed in 0.48.0.

Reviewer 3: My expectation is that `enableMetrics` will be replaced by:
```yaml
  metricsConfig:
    type: jmxPrometheusExporter
```

But that is missing from this proposal and it is not clear if that is intentional or unintentional.

> The `enableMetrics` property in `KafkaBridge` CRD will be deprecated in 0.46.0 and removed in 0.48.0.

Author: Yes, I will make it explicit.

The *Exporter* user provided configuration file will be stored in a ConfigMap and mounted in the Bridge's container.
The file path passed to the Bridge's container using the `bridge.metrics.exporter.config.path` property will be `/opt/strimzi/custom-config/metrics-config.yml`.

Metrics will be exposed through the Kafka Bridge's HTTP server, so the listener of the *Reporter* will be disabled.
Other configurations will be locked down with the exception of the `prometheus.metrics.reporter.allowlist` property.

## Affected/not affected projects

The affected projects are Cluster Operator and Kafka Bridge.

## Compatibility

### Bridge

The `KAFKA_BRIDGE_METRICS_ENABLED` environment variable will be deprecated in the next release, and removed in the first major or minor release of 2026.
When set, the user will get a warning suggesting to use the `bridge.metrics` property.
In case they are both set, `bridge.metrics` will take precedence over `KAFKA_BRIDGE_METRICS_ENABLED`.

#### Migration

When `KAFKA_BRIDGE_METRICS_ENABLED` will be removed, *Exporter* users will have to set `bridge.metrics=jmxPrometheusExporter` in their `application.properties`.

### Cluster Operator

> The `enableMetrics` property in `KafkaBridge` CRD will be deprecated in the next release and removed with the release of Strimzi v1 API.

Reviewer 5: If this proposal is accepted we should make sure to raise an issue to remove the `enableMetrics` property when we release the Strimzi v1 API so this isn't forgotten

Author: The `enableMetrics` property will be marked as deprecated, so I guess it will be automatically eligible for removal in v1.

Reviewer 3: Yeah - whoever works on the `v1` API would be expected to remove all the deprecated APIs.

When set, the user will get a warning suggesting to use the `metricsConfig` configuration.
In case they are both set, `metricsConfig` will take precedence over `enableMetrics`.

#### Migration

> When `enableMetrics` will be removed, *Exporter* users will have to create a secret with its configuration, and reference the secret as shown in the following example:

Reviewer 3: ```suggestion
When `enableMetrics` will be removed, *Exporter* users will have to create a config map with its configuration, and reference it as shown in the following example:
```

```yaml
spec:
  metricsConfig:
    type: jmxPrometheusExporter
    valueFrom:
      configMapKeyRef:
        name: my-metrics
        key: my-metrics-config.yml
```

> A full example that include the *Exporter* configuration secret will be provided.

Reviewer 3: ```suggestion
A full example that include the *Exporter* configuration config map will be provided.
```
